flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveContinuousMonitoringFunction.java

```diff
@@ -0,0 +1,280 @@
+			long startupTimestampMillis,
+			PartitionStrategyFactory partStrategyFactory,
+			int readerParallelism,
+			long interval) {
+		this.hiveShim = hiveShim;
+		this.conf = new JobConfWrapper(conf);
+		this.tablePath = tablePath;
+		this.partitionKeys = catalogTable.getPartitionKeys();
+		this.fieldNames = catalogTable.getSchema().getFieldNames();
+		this.fieldTypes = catalogTable.getSchema().getFieldDataTypes();
+		this.startupTimestampMillis = startupTimestampMillis;
+		this.partStrategyFactory = partStrategyFactory;
+
+		this.interval = interval;
+		this.readerParallelism = Math.max(readerParallelism, 1);
+		this.currentReadTime = Long.MIN_VALUE;
+	}
+
+	@Override
+	public void initializeState(FunctionInitializationContext context) throws Exception {
+		this.strategy = partStrategyFactory.createStrategy(getRuntimeContext().getUserCodeClassLoader());
+
+		this.currReadTimeState = context.getOperatorStateStore().getListState(
+			new ListStateDescriptor<>(
+				"partition-monitoring-state",
+				LongSerializer.INSTANCE
+			)
+		);
+		this.distinctPartsState = context.getOperatorStateStore().getListState(
+			new ListStateDescriptor<>(
+				"partition-monitoring-state",
+				new ListSerializer<>(new ListSerializer<>(StringSerializer.INSTANCE))
+			)
+		);
+
+		this.client = this.hiveShim.getHiveMetastoreClient(new HiveConf(conf.conf(), HiveConf.class));
+
+		Table hiveTable = client.getTable(tablePath.getDatabaseName(), tablePath.getObjectName());
+		this.tableProps = HiveReflectionUtils.getTableMetadata(hiveShim, hiveTable);
+		this.defaultPartitionName = conf.conf().get(HiveConf.ConfVars.DEFAULTPARTITIONNAME.varname,
+				HiveConf.ConfVars.DEFAULTPARTITIONNAME.defaultStrVal);
+
+		this.distinctPartitions = new HashSet<>();
+		if (context.isRestored()) {
+			LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+			this.currentReadTime = this.currReadTimeState.get().iterator().next();
+			this.distinctPartitions.addAll(this.distinctPartsState.get().iterator().next());
+		} else {
+			LOG.info("No state to restore for the {}.", getClass().getSimpleName());
+			this.currentReadTime = this.startupTimestampMillis;
+		}
+	}
+
+	@Override
+	public void run(SourceContext<TimestampedHiveInputSplit> context) throws Exception {
+		checkpointLock = context.getCheckpointLock();
+		while (isRunning) {
+			synchronized (checkpointLock) {
+				monitorAndForwardSplits(context);
+			}
+			Thread.sleep(interval);
+		}
+	}
+
+	private void monitorAndForwardSplits(
+			SourceContext<TimestampedHiveInputSplit> context) throws IOException, TException {
+		assert (Thread.holdsLock(checkpointLock));
+
+		List<Partition> partitions = client.listPartitionsByFilter(
+				tablePath.getDatabaseName(),
+				tablePath.getObjectName(),
+				strategy.generateFetchFilter(partitionKeys, currentReadTime),
+				(short) -1);
+
+		if (partitions.isEmpty()) {
+			return;
+		}
+
+		long maxTime = Long.MIN_VALUE;
+		for (Partition partition : partitions) {
```

Comment:
The java doc mentions the splits should be forwarded in a specific order. But it seems we're not sorting the partitions here?

Reply:
Good point, yes, will sort.